Feature/structured logging

dfurtado opened the pull request:
Added implementation of a `LoggerContext` which provides support for structured logs. Along with the implementation of the `LoggerContext` it has been added to the `Logger` struct a field to receive a formatter that will be used when the log entry is being sent to the configured output.

When using the `LoggerContext` instead of the Logger, the message passed in the logging methods (Info, Warning, Error, Debug...) will be included in the as part of the JSON object in the key "msg" together with all the other fields specified when using the `WithContext` method.

dfurtado committed:
Initial implementation of structured log support

Added implementation of a LoggerContext which provides support for
structured logs. Along with the implementation of the LoggerContext it
has been added to the Logger struct a field to receive a formatter
that will be used when the log entry is being sent to the configured
output.

When using the LoggerContext instead of the Logger, the message passed
in the logging methods (Info, Warning, Error, Debug...) will be included in
the as part of the JSON object in the key "msg" together with all the other
fields specified when using the WithContext method.

dfurtado committed:
Reverting docs wrongly modified.

dfurtado committed:
Reverting doc wrongly modified.

dfurtado committed:
Updated documentation

Updated documentation with code snippets showing how to create
structured logs.

dfurtado committed:
Skip serializing fields depending of wlog.LogLevel

If the current logLevel is less than the level configured on the logger
we skip all the field serialization.

ConsulFirmin commented:
Looks promising!  But with that said, I got a few comments :-)

## Linter
Run go linter and fix comments. 
Ex 1. exported type Fields should have comment or be unexported
Ex 2. comment on exported method LoggerContext.Debug should be of the form "Debug ..."

## Non-idiomatic use of  WithContext
WithContext() should only be used with context.Context in idiomatic Go. The same goes for ctx variable name.. Please rename or utilize real context.Contexts to hold context-bound data. I'm not sure if context.Context is a good idea here though. Could use context.WithValue() I suppose.

## All base properties should be included in the json-formatted output
Formatting output should include common properties timestamp and log-level

## Inconsistency when configuring formatter
Currently you configure a formatter on the logger-instance but it's only taken into account when using a LoggerContext (e.g. wlog.Info(..) will output plain text even if SetFormatter is configured for JSON). That doesn't make sense and needs to be addressed. Either remove the SetFormatter (for now) from wlog.Logger and implictly use the JSONFormatter for the LoggerContext or make sure SetFormatter is honored for both wlog and LoggerContext.

##  Code cleanup (?)
Two instances of LoggerContext gets created and allocation of 6 entries in map is hard-coded?

// WithContext returns a new instance of LoggerContext
func (l *Logger) WithContext(f Fields) *LoggerContext {
	newContext := &LoggerContext{logger: l, fields: make(Fields, 6)}
	return newContext.WithContext(f)
}

Why not just:

// WithContext returns a new instance of LoggerContext
func (l *Logger) WithContext(f Fields) *LoggerContext {
	return &LoggerContext{l, f}
}

## Config struct should be updated with support for formatters
wlog.Config should be updated to support a default formatter (low priority, nevermind for now) 

## Tests!
Add test for formatters!

## Stuff to consider later
When this PR is merged, we should consider introducing a TextFormatter or similar for the non-JSON ouput and move the current text-only writer to that formatter. But let's worry about that later.

dfurtado committed:
Changes after code review

- Renamed WithContext to WithFields to make it clear that is not go
Contexts.
- Added log level and timestamp in the JSON output.
- JSONFormatter is added automatically to the LoggerContext when
WithFields is called.
- Removed SetFormatter from wlog.
- Fixed docstrings on exported methods.
- Changed wlog.Write to wlog.write since it is used only internally.
- Update README with the new changes.

dfurtado committed:
docstring fixes.

dfurtado committed:
Added tests and minor fixes

dfurtado committed:
Code formatting

dfurtado committed:
Refactoring: removing LoggerContext

- Removed the ĹoggerContext and moved some of its methods to Logger
- Updated in the formatters and added a Textformatter which is going to
be the default formatter when a new instance of Logger is created
- Added unit tests for JSONFormatter

dfurtado committed:
Updated README

Updated README with documentation about structured logs.

dfurtado committed:
Code cleanup

dfurtado committed:
Removed buffer from Logger

- removed buffer from Logger since we are using a pool of buffers
instead
- Moved getTimestamp function to the bottom with the file for a better
organization of the code.
